<a href="https://colab.research.google.com/github/GuilleArditti/Introduccion-al-Analisis-de-Datos---Alkemy/blob/main/ProyectoFinal.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# *Proyecto Final*

## *Introducción al Análisis de Datos - Alkemy*

## *Alumno: Guillermo Arditti*






## **Setup**

Importación de la libreria Pandas.

In [277]:
!pip install pandas

In [278]:
import pandas as pd
import os


## **CARGA DE CSV´S**

A continuación se cargan los CSV´s a trabajar para el proyecto


### Customers

In [279]:
df_customers = pd.read_csv("https://raw.githubusercontent.com/GuilleArditti/Introduccion-al-Analisis-de-Datos---Alkemy/main/Datasets-Originales/ecommerce_customers_dataset.csv")
df_customers.head() #Leo las primeras 5 lineas para verificar funcionamiento

customer_id                customer_unique_id  \
0  06b8999e2fba1a1fbc88172c00ba8bc7  861eff4711a542e4b93843c6dd7febb0   
1  18955e83d337fd6b2def6b18a428ac77  290c77bc529b7ac935b93aa66c333dc3   
2  4e7b3e00288586ebd08712fdd0374a03  060e732b5b29e8181a18229c7b0b2b5e   
3  b2b6027bc5c5109e529d4dc6358b12c3  259dac757896d24d7702b9acbbff3f3c   
4  4f2d8ab171c80ec8364f7c12e35b23ad  345ecd01c38d18a9036ed96c73b8d066   

   customer_zip_code_prefix          customer_city customer_state  
0                     14409                 franca             SP  
1                      9790  sao bernardo do campo             SP  
2                      1151              sao paulo             SP  
3                      8775        mogi das cruzes             SP  
4                     13056               campinas             SP

### Order_items

In [280]:
df_order_items = pd.read_csv("https://raw.githubusercontent.com/GuilleArditti/Introduccion-al-Analisis-de-Datos---Alkemy/main/Datasets-Originales/ecommerce_order_items_dataset.csv")
df_order_items.head() #Leo las primeras 5 lineas para verificar funcionamiento

order_id  order_item_id  \
0  00010242fe8c5a6d1ba2dd792cb16214              1   
1  00018f77f2f0320c557190d7a144bdd3              1   
2  000229ec398224ef6ca0657da4fc703e              1   
3  00024acbcdf0a6daa1e931b038114c75              1   
4  00042b26cf59d7ce69dfabb4e55b4fd9              1   

                         product_id                         seller_id  \
0  4244733e06e7ecb4970a6e2683c13e61  48436dade18ac8b2bce089ec2a041202   
1  e5f2d52b802189ee658865ca93d83a8f  dd7ddc04e1b6c2c614352b383efe2d36   
2  c777355d18b72b67abbeef9df44fd0fd  5b51032eddd242adc84c38acab88f23d   
3  7634da152a4610f1595efa32f14722fc  9d7a1d34a5052409006425275ba1c2b4   
4  ac6c3623068f30de03045865e4e10089  df560393f3a51e74553ab94004ba5c87   

   shipping_limit_date   price  freight_value  
0  2017-09-19 09:45:35   58.90          13.29  
1  2017-05-03 11:05:13  239.90          19.93  
2  2018-01-18 14:48:30  199.00          17.87  
3  2018-08-15 10:10:18   12.99          12.79  
4  2017-02-13 13:57:51  199.90          18.14

### Order_payments

In [281]:
df_order_payments = pd.read_csv("https://raw.githubusercontent.com/GuilleArditti/Introduccion-al-Analisis-de-Datos---Alkemy/main/Datasets-Originales/ecommerce_order_payments_dataset.csv")
df_order_payments.head() #Leo las primeras 5 lineas para verificar funcionamiento

order_id  payment_sequential payment_type  \
0  b81ef226f3fe1789b1e8b2acac839d17                   1  credit_card   
1  a9810da82917af2d9aefd1278f1dcfa0                   1  credit_card   
2  25e8ea4e93396b6fa0d3dd708e76c1bd                   1  credit_card   
3  ba78997921bbcdc1373bb41e913ab953                   1  credit_card   
4  42fdf880ba16b47b59251dd489d4441a                   1  credit_card   

   payment_installments  payment_value  
0                     8          99.33  
1                     1          24.39  
2                     1          65.71  
3                     8         107.78  
4                     2         128.45

### Orders

In [282]:
df_orders = pd.read_csv("https://raw.githubusercontent.com/GuilleArditti/Introduccion-al-Analisis-de-Datos---Alkemy/main/Datasets-Originales/ecommerce_orders_dataset.csv")
df_orders.head() #Leo las primeras 5 lineas para verificar funcionamiento

order_id                       customer_id  \
0  e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
1  53cdb2fc8bc7dce0b6741e2150273451  b0830fb4747a6c6d20dea0b8c802d7ef   
2  47770eb9100c2d0c44946d9cf07ec65d  41ce2a54c0b03bf3443c3d931a367089   
3  949d5b44dbf5de918fe9c16f97b45f8a  f88197465ea7920adcdbec7375364d82   
4  ad21c59c0840e6cb83a9ceb5573f8159  8ab97904e6daea8866dbdbc4fb7aad2c   

  order_status order_purchase_timestamp    order_approved_at  \
0    delivered      2017-10-02 10:56:33  2017-10-02 11:07:15   
1    delivered      2018-07-24 20:41:37  2018-07-26 03:24:27   
2    delivered      2018-08-08 08:38:49  2018-08-08 08:55:23   
3    delivered      2017-11-18 19:28:06  2017-11-18 19:45:59   
4    delivered      2018-02-13 21:18:39  2018-02-13 22:20:29   

  order_delivered_carrier_date order_delivered_customer_date  \
0          2017-10-04 19:55:00           2017-10-10 21:25:13   
1          2018-07-26 14:31:00           2018-08-07 15:27:45   
2          2018-08-08 13:50:00           2018-08-17 18:06:29   
3          2017-11-22 13:39:59           2017-12-02 00:28:42   
4          2018-02-14 19:46:34           2018-02-16 18:17:02   

  order_estimated_delivery_date  
0           2017-10-18 00:00:00  
1           2018-08-13 00:00:00  
2           2018-09-04 00:00:00  
3           2017-12-15 00:00:00  
4           2018-02-26 00:00:00

### Products

In [283]:
df_products = pd.read_csv("https://raw.githubusercontent.com/GuilleArditti/Introduccion-al-Analisis-de-Datos---Alkemy/main/Datasets-Originales/ecommerce_products_dataset.csv")
df_products.head() #Leo las primeras 5 lineas para verificar funcionamiento

product_id  product_category_name  \
0  1e9e8ef04dbcff4541ed26657ea517e5             perfumaria   
1  3aa071139cb16b67ca9e5dea641aaa2f                  artes   
2  96bd76ec8810374ed1b65e291975717f          esporte_lazer   
3  cef67bcfe19066a932b7673e239eb23d                  bebes   
4  9dc1a7de274444849c219cff195d0b71  utilidades_domesticas   

   product_name_lenght  product_description_lenght  product_photos_qty  \
0                 40.0                       287.0                 1.0   
1                 44.0                       276.0                 1.0   
2                 46.0                       250.0                 1.0   
3                 27.0                       261.0                 1.0   
4                 37.0                       402.0                 4.0   

   product_weight_g  product_length_cm  product_height_cm  product_width_cm  
0             225.0               16.0               10.0              14.0  
1            1000.0               30.0               18.0              20.0  
2             154.0               18.0                9.0              15.0  
3             371.0               26.0                4.0              26.0  
4             625.0               20.0               17.0              13.0

## **VISUALIZACIÓN DE DATOS DE LOS DATAFRAMES**

Se visualizarán los datos de los DataFrames cargados para saber sobre su tamaño, su cantidad de nulos (si es que tienen), que tipos de datos los conforman y si tienen o no, valores duplicados.

### Funciones de Visualización

Se aplicaran distintas funciones para la visualización de los DataFrames.

#### Función de Tamaño

Función que muestra el tamaño del DataFrame pasado como argumento.

In [284]:
def tamañoDataFrame(df):
  print("---------------------")
  print("")
  print("Tamaño del DataFrame:")
  print("Cantidad de filas:", df.shape[0])
  print("Cantidad de columnas:", df.shape[1])
  print("")
  print("---------------------")


#### Función Cantidad de Nulos

Función que muestra la cantidad de nulos del DataFrame pasado como argumento.

In [285]:
def cantidadDeNulos(df):
  print("---------------------")
  print("")
  print("Cantidad de valores nulos por columna:")
  print(df.isnull().sum())
  print("")
  print("Cantidad de valores nulos en total: ", df.isnull().sum().sum())
  print("---------------------")

#### Función Tipos de Datos

Función que muestra los distintos tipos de datos que contiene el DataFrame pasado como argumento.

In [286]:
def tiposDeDatos(df):
  print("---------------------")
  print("")
  print("Tipos de datos del DataFrame:")
  print(df.info())
  print("")
  print("---------------------")

#### Función Cantidad de Duplicados

Función que muestra la cantidad de valores duplicados que tiene el Dataframe pasado como argumento.

In [287]:
def cantidadDeDuplicados(df):
  print("---------------------")
  print("")
  print("Cantidad de valores duplicados del DataFrame:")
  print(df.duplicated().sum())
  print("")
  print("---------------------")

### Implementación de las Funciones de Visualización en los DataFrames

---



#### Customers

In [288]:
tamañoDataFrame(df_customers)
cantidadDeNulos(df_customers)
tiposDeDatos(df_customers)
cantidadDeDuplicados(df_customers)

---------------------

Tamaño del DataFrame:
Cantidad de filas: 99441
Cantidad de columnas: 5

---------------------
---------------------

Cantidad de valores nulos por columna:
customer_id                 0
customer_unique_id          0
customer_zip_code_prefix    0
customer_city               0
customer_state              0
dtype: int64

Cantidad de valores nulos en total:  0
---------------------
---------------------

Tipos de datos del DataFrame:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 99441 entries, 0 to 99440
Data columns (total 5 columns):
 #   Column                    Non-Null Count  Dtype 
---  ------                    --------------  ----- 
 0   customer_id               99441 non-null  object
 1   customer_unique_id        99441 non-null  object
 2   customer_zip_code_prefix  99441 non-null  int64 
 3   customer_city             99441 non-null  object
 4   customer_state            99441 non-null  object
dtypes: int64(1), object(4)
memory usage: 3.8+ MB
None

-

#### Order_items

In [289]:
tamañoDataFrame(df_order_items)
cantidadDeNulos(df_order_items)
tiposDeDatos(df_order_items)
cantidadDeDuplicados(df_order_items)

---------------------

Tamaño del DataFrame:
Cantidad de filas: 112650
Cantidad de columnas: 7

---------------------
---------------------

Cantidad de valores nulos por columna:
order_id               0
order_item_id          0
product_id             0
seller_id              0
shipping_limit_date    0
price                  0
freight_value          0
dtype: int64

Cantidad de valores nulos en total:  0
---------------------
---------------------

Tipos de datos del DataFrame:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 112650 entries, 0 to 112649
Data columns (total 7 columns):
 #   Column               Non-Null Count   Dtype  
---  ------               --------------   -----  
 0   order_id             112650 non-null  object 
 1   order_item_id        112650 non-null  int64  
 2   product_id           112650 non-null  object 
 3   seller_id            112650 non-null  object 
 4   shipping_limit_date  112650 non-null  object 
 5   price                112650 non-null  float64

#### Order_payments

In [290]:
tamañoDataFrame(df_order_payments)
cantidadDeNulos(df_order_payments)
tiposDeDatos(df_order_payments)
cantidadDeDuplicados(df_order_payments)

---------------------

Tamaño del DataFrame:
Cantidad de filas: 103886
Cantidad de columnas: 5

---------------------
---------------------

Cantidad de valores nulos por columna:
order_id                0
payment_sequential      0
payment_type            0
payment_installments    0
payment_value           0
dtype: int64

Cantidad de valores nulos en total:  0
---------------------
---------------------

Tipos de datos del DataFrame:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 103886 entries, 0 to 103885
Data columns (total 5 columns):
 #   Column                Non-Null Count   Dtype  
---  ------                --------------   -----  
 0   order_id              103886 non-null  object 
 1   payment_sequential    103886 non-null  int64  
 2   payment_type          103886 non-null  object 
 3   payment_installments  103886 non-null  int64  
 4   payment_value         103886 non-null  float64
dtypes: float64(1), int64(2), object(2)
memory usage: 4.0+ MB
None

--------------------

#### Orders

In [291]:
tamañoDataFrame(df_orders)
cantidadDeNulos(df_orders)
tiposDeDatos(df_orders)
cantidadDeDuplicados(df_orders)

---------------------

Tamaño del DataFrame:
Cantidad de filas: 99441
Cantidad de columnas: 8

---------------------
---------------------

Cantidad de valores nulos por columna:
order_id                            0
customer_id                         0
order_status                        0
order_purchase_timestamp            0
order_approved_at                 160
order_delivered_carrier_date     1783
order_delivered_customer_date    2965
order_estimated_delivery_date       0
dtype: int64

Cantidad de valores nulos en total:  4908
---------------------
---------------------

Tipos de datos del DataFrame:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 99441 entries, 0 to 99440
Data columns (total 8 columns):
 #   Column                         Non-Null Count  Dtype 
---  ------                         --------------  ----- 
 0   order_id                       99441 non-null  object
 1   customer_id                    99441 non-null  object
 2   order_status                   99441 

#### Products

In [292]:
tamañoDataFrame(df_products)
cantidadDeNulos(df_products)
tiposDeDatos(df_products)
cantidadDeDuplicados(df_products)

---------------------

Tamaño del DataFrame:
Cantidad de filas: 32951
Cantidad de columnas: 9

---------------------
---------------------

Cantidad de valores nulos por columna:
product_id                      0
product_category_name         610
product_name_lenght           610
product_description_lenght    610
product_photos_qty            610
product_weight_g                2
product_length_cm               2
product_height_cm               2
product_width_cm                2
dtype: int64

Cantidad de valores nulos en total:  2448
---------------------
---------------------

Tipos de datos del DataFrame:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32951 entries, 0 to 32950
Data columns (total 9 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   product_id                  32951 non-null  object 
 1   product_category_name       32341 non-null  object 
 2   product_name_lenght         32341 non-null 

## **CONVERSION DE DATAFRAMES**

Se manipularán los DataFrames con el objetivo de que los datos que contienen sean homogéneos, garantizando una correcta lectura de los mismos.

### Funciones de Tratamiento

Funciones que se utilizarán para la limpieza y procesamiento de datos.

#### Función convertirAMayúsculas

Función que como parámetros un DataFrame y una columna del mismo, para pasar lo contenido en dicha columna a mayúsculas.

In [293]:
def convertirAMayusculas(df, columna):
    df[columna] = df[columna].str.upper()
    return df

#### Función convertirAMinúsculas

Función que toma como parámetros un DataFrame y una columna del mismo, para pasar lo contenido en dicha columna a minúsculas.

In [294]:
def convertirAMinusculas(df, columna):
    df[columna] = df[columna].str.lower()
    return df

#### Función eliminarNulos

Función que toma como parámetros un DataFrame y un porcentaje límite dado. Si el porcentaje de nulos de una columna es menor al porcentaje límite (n), elimina dicha fila en la columna.

In [295]:
def eliminarNulos(df,n):
  porcentaje_nulos_df=(df.isnull().sum()/len(df))*100

  #Columnas que cumplen que su porcentaje es menor a n
  columnas_interesantes=porcentaje_nulos_df[porcentaje_nulos_df<=n].index

  #Se eliminan las filas en las columnas seleccionadas
  df.dropna(subset=columnas_interesantes, inplace=True)

  print("Porcentaje de nulos del Dataframe por columna: ")
  print(porcentaje_nulos_df)

#### Función transformarAFecha

Función que toma como parámetros a un Dataframe y a las columnas del mismo que se pasarán a formato fecha.

In [296]:
def transformarAFecha(df, columnas_fechas):
    df[columnas_fechas] = df[columnas_fechas].apply(pd.to_datetime)
    return df

#### Función transformarAString

Función que toma como parámetro a un dataframe y a un conjunto de columnas, luego toma cada una de ellas y convierte su contenido a formato String

In [297]:
def transformarAString(df, columnas_string):
    for columna in columnas_string:
        df[columna] = df[columna].astype("string")
    return df

### Implementación de las funciones de Tratamiento en los DataFrames

De acuerdo a lo visto en el apartado "Análisis de Datos de los DataFrames", podemos concluir que debemos realizar las siguientes operaciones en los DataFrames:




#### Customers



In [298]:
#Columnas a transfomar en string
columnas_a_string = ["customer_id","customer_unique_id","customer_zip_code_prefix","customer_city","customer_state"]

df_customers = transformarAString(df_customers, columnas_a_string)

#Columnas a transformar a Mayúsculas
df_customers = convertirAMayusculas(df_customers, "customer_state")

#Columnas a transformar a Minúsculas
df_customers = convertirAMinusculas(df_customers, "customer_city")

#Establecemos la clave primaria
#establecerClavePrimaria(df_customers)

#Mostramos el resultado
df_customers.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 99441 entries, 0 to 99440
Data columns (total 5 columns):
 #   Column                    Non-Null Count  Dtype 
---  ------                    --------------  ----- 
 0   customer_id               99441 non-null  string
 1   customer_unique_id        99441 non-null  string
 2   customer_zip_code_prefix  99441 non-null  string
 3   customer_city             99441 non-null  string
 4   customer_state            99441 non-null  string
dtypes: string(5)
memory usage: 3.8 MB


#### Order_items

In [299]:
#Columnas que contienen fechas
columnas_fechas_order_items = ["shipping_limit_date"]

df_order_items = transformarAFecha(df_order_items, columnas_fechas_order_items)

#Columnas a transfomar en string
columnas_string_order_items = ["order_id","order_item_id","product_id","seller_id"]

df_order_items = transformarAString(df_order_items, columnas_string_order_items)

#Mostramos el resultado
df_order_items.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 112650 entries, 0 to 112649
Data columns (total 7 columns):
 #   Column               Non-Null Count   Dtype         
---  ------               --------------   -----         
 0   order_id             112650 non-null  string        
 1   order_item_id        112650 non-null  string        
 2   product_id           112650 non-null  string        
 3   seller_id            112650 non-null  string        
 4   shipping_limit_date  112650 non-null  datetime64[ns]
 5   price                112650 non-null  float64       
 6   freight_value        112650 non-null  float64       
dtypes: datetime64[ns](1), float64(2), string(4)
memory usage: 6.0 MB


#### Order_payments

In [300]:
#Conversión a minusculas de la culumna "payment_type"
df_order_payments = convertirAMinusculas(df_order_payments, "payment_type")

#Columnas a transfomar en string
columnas_string_order_payments = ["order_id","payment_type"]

df_order_payments = transformarAString(df_order_payments, columnas_string_order_payments)

#Mostramos el resultado
df_order_payments.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 103886 entries, 0 to 103885
Data columns (total 5 columns):
 #   Column                Non-Null Count   Dtype  
---  ------                --------------   -----  
 0   order_id              103886 non-null  string 
 1   payment_sequential    103886 non-null  int64  
 2   payment_type          103886 non-null  string 
 3   payment_installments  103886 non-null  int64  
 4   payment_value         103886 non-null  float64
dtypes: float64(1), int64(2), string(2)
memory usage: 4.0 MB


#### Orders

In [301]:
#Eliminamos los valores nulos (Importante: el primer run muestra el dataframe aún con nulos debido al print de la función, al correr de nuevo se muestra correctamente)
eliminarNulos(df_orders,10) #En este caso, determinamos al valor límite de porcentajes de nulos al 10%.

#Columnas que contienen fechas
columnas_fechas_orders = ['order_purchase_timestamp', 'order_approved_at',
                          'order_delivered_carrier_date', 'order_delivered_customer_date',
                          'order_estimated_delivery_date']

df_orders = transformarAFecha(df_orders, columnas_fechas_orders)

#Columnas a transformar en string
columnas_string_orders = ["order_id","customer_id","order_status"]

df_orders = transformarAString(df_orders, columnas_string_orders)

#Mostramos el resultado
df_orders.info()

Porcentaje de nulos del Dataframe por columna: 
order_id                         0.000000
customer_id                      0.000000
order_status                     0.000000
order_purchase_timestamp         0.000000
order_approved_at                0.160899
order_delivered_carrier_date     1.793023
order_delivered_customer_date    2.981668
order_estimated_delivery_date    0.000000
dtype: float64
<class 'pandas.core.frame.DataFrame'>
Index: 96461 entries, 0 to 99440
Data columns (total 8 columns):
 #   Column                         Non-Null Count  Dtype         
---  ------                         --------------  -----         
 0   order_id                       96461 non-null  string        
 1   customer_id                    96461 non-null  string        
 2   order_status                   96461 non-null  string        
 3   order_purchase_timestamp       96461 non-null  datetime64[ns]
 4   order_approved_at              96461 non-null  datetime64[ns]
 5   order_delivered_carrier_

#### Products

In [302]:
#Columnas a transformar en string
columnas_string_products = ["product_id","product_category_name"]

df_products = transformarAString(df_products, columnas_string_products)

#Conversion a minúsculas de la columna "product_category_name"
df_products = convertirAMinusculas(df_products, "product_category_name")

#Eliminamos nulos (Importante: el primer run muestra el dataframe aún con nulos debido al print de la función, al correr de nuevo se muestra correctamente)
eliminarNulos(df_products,10) #En este caso, determinamos al valor límite de porcentajes de nulos al 10%.

#Mostramos el resultado
df_products.info()

Porcentaje de nulos del Dataframe por columna: 
product_id                    0.000000
product_category_name         1.851234
product_name_lenght           1.851234
product_description_lenght    1.851234
product_photos_qty            1.851234
product_weight_g              0.006070
product_length_cm             0.006070
product_height_cm             0.006070
product_width_cm              0.006070
dtype: float64
<class 'pandas.core.frame.DataFrame'>
Index: 32340 entries, 0 to 32950
Data columns (total 9 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   product_id                  32340 non-null  string 
 1   product_category_name       32340 non-null  string 
 2   product_name_lenght         32340 non-null  float64
 3   product_description_lenght  32340 non-null  float64
 4   product_photos_qty          32340 non-null  float64
 5   product_weight_g            32340 non-null  float64
 6   product_length_cm     

## **ANALISIS DE DATOS**

Luego de convertir los DataFrames, realizamos un análisis de los datos que contienen, para poder mostrar insights.

### Función AnalizarDataFrame

In [303]:
def analizarDataFrame(df):
  columnas_numericas = df.select_dtypes(include=['int', 'float']).columns
  if not columnas_numericas.empty:
    print("valores numéricos: ")
    print(df[columnas_numericas].describe())
    print("-------------------")
  columnas_string= df.select_dtypes(include=["string"]).columns
  if not columnas_string.empty:
    print("valores descriptivos:")
    print(df[columnas_string].describe())
    print("-------------------")
  columnas_fechas=df.select_dtypes(include=["datetime"]).columns
  if not columnas_fechas.empty:
    print("valores fechas:")
    print(df[columnas_fechas].describe())
    print("-------------------")

#### Customers


In [304]:
#Para conocer la cantidad de clientes únicos del conjunto de datos Customers
print("Cantidad de clientes únicos: ", df_customers["customer_unique_id"].nunique())

Cantidad de clientes únicos:  96096


In [305]:
analizarDataFrame(df_customers)

valores descriptivos:
                             customer_id                customer_unique_id  \
count                              99441                             99441   
unique                             99441                             96096   
top     06b8999e2fba1a1fbc88172c00ba8bc7  8d50f5eadf50201ccdcedfb9e2ac8455   
freq                                   1                                17   

       customer_zip_code_prefix customer_city customer_state  
count                     99441         99441          99441  
unique                    14994          4119             27  
top                       22790     sao paulo             SP  
freq                        142         15540          41746  
-------------------


In [306]:
#Realizamos un analisis por ciudad de los compradores
df_customers["customer_city"].value_counts()

customer_city
sao paulo            15540
rio de janeiro        6882
belo horizonte        2773
brasilia              2131
curitiba              1521
                     ...  
palmares paulista        1
pires ferreira           1
vicente dutra            1
alagoa nova              1
eugenio de castro        1
Name: count, Length: 4119, dtype: Int64

Como se puede observar, el mayor caudal de compradores se encuentra en Sao Paulo, Rio de Janerio, Belo Horizonte, Brasilia y Curitiba, mientras que en ciudades como Alagao Nova o Eugenio de Castro solo hay un comprador en cada una.

####Order_items

In [307]:
analizarDataFrame(df_order_items)

valores numéricos: 
               price  freight_value
count  112650.000000  112650.000000
mean      120.653739      19.990320
std       183.633928      15.806405
min         0.850000       0.000000
25%        39.900000      13.080000
50%        74.990000      16.260000
75%       134.900000      21.150000
max      6735.000000     409.680000
-------------------
valores descriptivos:
                                order_id order_item_id  \
count                             112650        112650   
unique                             98666            21   
top     8272b63d03f5f79c56e9e4120aec44ef             1   
freq                                  21         98666   

                              product_id                         seller_id  
count                             112650                            112650  
unique                             32951                              3095  
top     aca2eb7d00ea1a7b8ebd4e68314663af  6560211a19b47992c3666cc44a7e94c0  
freq           

In [308]:
print(df_order_items["seller_id"].value_counts())

seller_id
6560211a19b47992c3666cc44a7e94c0    2033
4a3ca9315b744ce9f8e9374361493884    1987
1f50f920176fa81dab994f9023523100    1931
cc419e0650a3c5ba77189a1882b7556a    1775
da8622b14eb17ae2831f4ac5b9dab84a    1551
                                    ... 
3289d8c3eef4873f91ce0b1343c0dd58       1
f3055483058c84d75ca2a3314fb4deef       1
be67f78487e2cecb0d55bc769709e4f5       1
fedaedd3ca31d56ab33e92035e4b361c       1
13d95f0f6f73943d4ceffad0fc2cd32c       1
Name: count, Length: 3095, dtype: Int64


Aquí podemos ver que el vendedor que realizó mas ventas, realizó un total de 2033.




#### Order_payments

In [309]:
analizarDataFrame(df_order_payments)

valores numéricos: 
       payment_sequential  payment_installments  payment_value
count       103886.000000         103886.000000  103886.000000
mean             1.092679              2.853349     154.100380
std              0.706584              2.687051     217.494064
min              1.000000              0.000000       0.000000
25%              1.000000              1.000000      56.790000
50%              1.000000              1.000000     100.000000
75%              1.000000              4.000000     171.837500
max             29.000000             24.000000   13664.080000
-------------------
valores descriptivos:
                                order_id payment_type
count                             103886       103886
unique                             99440            5
top     fa65dad1b0e818e3ccc5cb0e39231352  credit_card
freq                                  29        76795
-------------------


In [310]:
print(df_order_payments["payment_value"].value_counts())

payment_value
50.00      324
20.00      274
100.00     255
77.57      250
35.00      165
          ... 
264.64       1
1071.83      1
563.95       1
38.07        1
363.31       1
Name: count, Length: 29077, dtype: int64


En este DataFrame podemos ver que la mayoría de pagos realizados por pedido no supera los 100 reales, siendo el pago mas común 50 reales y el promedio 154 reales.

#### Orders

In [311]:
analizarDataFrame(df_orders)

valores descriptivos:
                                order_id                       customer_id  \
count                              96461                             96461   
unique                             96461                             96461   
top     e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
freq                                   1                                 1   

       order_status  
count         96461  
unique            2  
top       delivered  
freq          96455  
-------------------
valores fechas:
            order_purchase_timestamp              order_approved_at  \
count                          96461                          96461   
mean   2018-01-01 23:53:26.642249216  2018-01-02 10:10:06.480142336   
min              2016-09-15 12:16:38            2016-09-15 12:16:38   
25%              2017-09-14 09:28:28            2017-09-14 14:30:14   
50%              2018-01-20 19:59:42            2018-01-22 13:49:00   
75%            

In [312]:
print(df_orders["order_delivered_carrier_date"].value_counts())

order_delivered_carrier_date
2018-05-09 15:48:00    47
2018-05-10 18:29:00    32
2018-05-07 12:31:00    21
2018-07-24 16:07:00    16
2018-05-02 15:15:00    16
                       ..
2017-10-27 20:07:58     1
2017-11-21 18:57:37     1
2018-03-21 20:03:37     1
2017-11-11 14:41:37     1
2018-03-09 22:11:59     1
Name: count, Length: 80100, dtype: int64


Como se puede observar, la mayor cantidad de ordenes tomadas por los transportistas para ser entregadas se dan luego del mediodía y en mayor proporción entre las 14:30 y las 17:00 horas.

In [313]:
#Cantidad total de pedidos realizados
print(df_orders["order_id"].nunique())

96461


#### Products

In [314]:
analizarDataFrame(df_products)

valores numéricos: 
       product_name_lenght  product_description_lenght  product_photos_qty  \
count         32340.000000                32340.000000        32340.000000   
mean             48.476592                  771.492393            2.188961   
std              10.245699                  635.124831            1.736787   
min               5.000000                    4.000000            1.000000   
25%              42.000000                  339.000000            1.000000   
50%              51.000000                  595.000000            1.000000   
75%              57.000000                  972.000000            3.000000   
max              76.000000                 3992.000000           20.000000   

       product_weight_g  product_length_cm  product_height_cm  \
count      32340.000000       32340.000000       32340.000000   
mean        2276.956586          30.854545          16.958813   
std         4279.291845          16.955965          13.636115   
min            0.

In [315]:
print(df_products["product_category_name"].value_counts())

product_category_name
cama_mesa_banho                  3029
esporte_lazer                    2867
moveis_decoracao                 2657
beleza_saude                     2444
utilidades_domesticas            2335
                                 ... 
fashion_roupa_infanto_juvenil       5
casa_conforto_2                     5
pc_gamer                            3
seguros_e_servicos                  2
cds_dvds_musicais                   1
Name: count, Length: 73, dtype: Int64


Aquí podemos observar que los productos que menos se venden son de la categoría: musical, seguros y servicios, artículos gamer, casa y confort, y ropa juvenil.

Por otro lado, la categoría de producto más vendida es la de cama, mesas y baño.

## GENERACION DE NUEVOS DATAFRAMES

Se generaran dos nuevos DataFrames a partir de los Dataframes originales, luego de haber sido sometidos a las conversiones que aplicamos previamente. Los exportaremos en formato csv para poder utilizarlos y generar un reporte con visualizaciones en Excel.

In [316]:
#Creamos una carpeta para guardar los csv´s a exportar
!mkdir dataset_generados

mkdir: cannot create directory ‘dataset_generados’: File exists


#### Order - Customers

In [317]:
#Realizamos un merge
df_merge_order_customers = pd.merge(df_orders, df_customers, how = "inner", on= "customer_id")
df_merge_order_customers.head()

#Quitamos aquellas columnas que no usaremos para el análisis visual
df_merge_order_customers.drop(['order_approved_at','customer_id','order_delivered_carrier_date','order_delivered_customer_date','order_estimated_delivery_date','customer_zip_code_prefix','customer_unique_id'], axis=1, inplace=True)

#Mostramos el resultado
df_merge_order_customers.head()

order_id order_status order_purchase_timestamp  \
0  e481f51cbdc54678b7cc49136f2d6af7    delivered      2017-10-02 10:56:33   
1  53cdb2fc8bc7dce0b6741e2150273451    delivered      2018-07-24 20:41:37   
2  47770eb9100c2d0c44946d9cf07ec65d    delivered      2018-08-08 08:38:49   
3  949d5b44dbf5de918fe9c16f97b45f8a    delivered      2017-11-18 19:28:06   
4  ad21c59c0840e6cb83a9ceb5573f8159    delivered      2018-02-13 21:18:39   

             customer_city customer_state  
0                sao paulo             SP  
1                barreiras             BA  
2               vianopolis             GO  
3  sao goncalo do amarante             RN  
4              santo andre             SP

In [318]:
#Exportamos como csv nuestro dataframe
df_merge_order_customers.to_csv("dataset_generados/order_customers.csv")

### Order items - Products

In [319]:
#Realizamos un merge
df_merge_order_items_products = pd.merge(df_order_items, df_products, how = "inner", on= "product_id")
df_merge_order_items_products.head()

#Quitamos aquellas columnas que no usaremos para el análisis visual
df_merge_order_items_products.drop(['product_id','seller_id','shipping_limit_date'], axis=1, inplace=True)

#Mostramos el resultado
df_merge_order_items_products.head()

order_id order_item_id  price  freight_value  \
0  00010242fe8c5a6d1ba2dd792cb16214             1   58.9          13.29   
1  130898c0987d1801452a8ed92a670612             1   55.9          17.96   
2  532ed5e14e24ae1f0d735b91524b98b9             1   64.9          18.33   
3  6f8c31653edb8c83e1a739408b5ff750             1   58.9          16.17   
4  7d19f4ef4d04461989632411b7e588b9             1   58.9          13.29   

  product_category_name  product_name_lenght  product_description_lenght  \
0            cool_stuff                 58.0                       598.0   
1            cool_stuff                 58.0                       598.0   
2            cool_stuff                 58.0                       598.0   
3            cool_stuff                 58.0                       598.0   
4            cool_stuff                 58.0                       598.0   

   product_photos_qty  product_weight_g  product_length_cm  product_height_cm  \
0                 4.0             650.0               28.0                9.0   
1                 4.0             650.0               28.0                9.0   
2                 4.0             650.0               28.0                9.0   
3                 4.0             650.0               28.0                9.0   
4                 4.0             650.0               28.0                9.0   

   product_width_cm  
0              14.0  
1              14.0  
2              14.0  
3              14.0  
4              14.0

In [320]:
#Exportamos como csv nuestro dataframe
df_merge_order_items_products.to_csv("dataset_generados/order_items_products.csv")